### ISSUE TITLE
SSSD does not handle group renames safely

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#1044]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/1040
- Created at 2011-10-14 13:23:52 by [sgallagh](https://pagure.io/user/sgallagh)
- Closed as Fixed
- Assigned to [jzeleny](https://pagure.io/user/jzeleny)
---

https://bugzilla.redhat.com/show_bug.cgi?id=746181

    Description of problem:
    We changed one of the groups name in MSAD that was defined as Unix group from
    uppercase to lowercase. After that we noticed problems with some of the
    services. In sssd_nss.log we got error messages:
    
    (Fri Oct 14 10:03:31 2011) [sssd[nss]] [nss_cmd_getgrgid_search] (0): getgrgid
    call returned more than one result !?!
    
    Problems:
    
    1. Checking groups in domain resulted in this:
    [root@SERVERNAME db]#  getent group |grep -i gr
    GROUPGROUP:*:10069:username
    groupgroup:*:10069:username
    
    2. Cache was not updated correctly even after we removed group from Domain and
    added it again.
    [root@SERVERNAME db]#  getent group |grep -i gr
    GROUPGROUP:*:10069:username
    groupgroup:*:10113:username
    
    We got expected result only after removing cache from /var/lib/sss/db
    directory. Resetting service didn't help.
    [root@SERVERNAME db]#  getent group |grep -i gr
    groupgroup:*:10113:username
    
    3. What is the configuration parameter responsible for cache refresh ? I would
    like groups / users and memberships to be updated asap.
    
    
    Why if MSAD is case insensitive it is important for SSSD what case group/user
    has ??
    
    
    
    Version-Release number of selected component (if applicable):
    sssd-client-1.5.1-34.el6_1.3.x86_64
    sssd-1.5.1-34.el6_1.3.x86_64

### Comments

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-10-14 13:49:44](https://pagure.io/SSSD/sssd/issue/1040#236670)**

This one is a little tricky. We store the group in the cache using the group's name as its DN. But if we change names and then do a lookup by GID, we don't purge the old group from the cache. So now it looks like we have two groups in the cache with the same ID.

Group name changes are tricky things. It's difficult to know whether a group's name has actually changed or we've looked up a new group with the same ID.

Since we can't really support the latter case anyway, I'm inclined to say that the fix for this would be to check the cache for other entries with the same ID before saving it and delete them if the name does not match.

This affects users as well.

coverity:  => 
description: https://bugzilla.redhat.com/show_bug.cgi?id=746181

{{{
Description of problem:
We changed one of the groups name in MSAD that was defined as Unix group from
uppercase to lowercase. After that we noticed problems with some of the
services. In sssd_nss.log we got error messages:

(Fri Oct 14 10:03:31 2011) [sssd[nss]] [nss_cmd_getgrgid_search] (0): getgrgid
call returned more than one result !?!

Problems:

1. Checking groups in domain resulted in this:
[root@SERVERNAME db]#  getent group |grep -i gr
GROUPGROUP:*:10069:username
groupgroup:*:10069:username

2. Cache was not updated correctly even after we removed group from Domain and
added it again.
[root@SERVERNAME db]#  getent group |grep -i gr
GROUPGROUP:*:10069:username
groupgroup:*:10113:username

We got expected result only after removing cache from /var/lib/sss/db
directory. Resetting service didn't help.
[root@SERVERNAME db]#  getent group |grep -i gr
groupgroup:*:10113:username

3. What is the configuration parameter responsible for cache refresh ? I would
like groups / users and memberships to be updated asap.

Why if MSAD is case insensitive it is important for SSSD what case group/user
has ??

Version-Release number of selected component (if applicable):
sssd-client-1.5.1-34.el6_1.3.x86_64
sssd-1.5.1-34.el6_1.3.x86_64
}}}
 => https://bugzilla.redhat.com/show_bug.cgi?id=746181

{{{
Description of problem:
We changed one of the groups name in MSAD that was defined as Unix group from
uppercase to lowercase. After that we noticed problems with some of the
services. In sssd_nss.log we got error messages:

(Fri Oct 14 10:03:31 2011) [sssd[nss]] [nss_cmd_getgrgid_search] (0): getgrgid
call returned more than one result !?!

Problems:

1. Checking groups in domain resulted in this:
[root@SERVERNAME db]#  getent group |grep -i gr
GROUPGROUP:*:10069:username
groupgroup:*:10069:username

2. Cache was not updated correctly even after we removed group from Domain and
added it again.
[root@SERVERNAME db]#  getent group |grep -i gr
GROUPGROUP:*:10069:username
groupgroup:*:10113:username

We got expected result only after removing cache from /var/lib/sss/db
directory. Resetting service didn't help.
[root@SERVERNAME db]#  getent group |grep -i gr
groupgroup:*:10113:username

3. What is the configuration parameter responsible for cache refresh ? I would
like groups / users and memberships to be updated asap.

Why if MSAD is case insensitive it is important for SSSD what case group/user
has ??

Version-Release number of selected component (if applicable):
sssd-client-1.5.1-34.el6_1.3.x86_64
sssd-1.5.1-34.el6_1.3.x86_64
}}}

patch:  => 0
rhbz:  => 
tests:  => 0
testsupdated:  => 0
upgrade:  => 0

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-10-20 14:59:09](https://pagure.io/SSSD/sssd/issue/1040#236671)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.7.0

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-11-02 15:30:43](https://pagure.io/SSSD/sssd/issue/1040#236672)**

Marking ticket #1076 a duplicate of this bug.

SSSD has a hard restriction in the SysDB that only one group can exist with a particular ID.

When a group is renamed on the server, it becomes an order-of-operations issue whether or not we handle it correctly. If we look up the old group name first, we'll see it's removed and delete it from our cache. If we ask for the new group name, we'll get an error attempting to save it to the cache because the IDs will match.

It's well-documented that we don't support multiple entries with the same GID and that doing so will result in unexpected behaviour, so I'd like to recommend the following change:

When we attempt to save a group to the sysdb and the lookup reveals that another group exists already with that GID, instead of failing we should remove the old group. If there actually ARE two groups with the same GID, the effect will be that they'll regularly remove each other, causing a lot more LDAP lookups (and confusing offline access), but this isn't a supported configuration anyway.

In the case that it's actually a rename, we'll handle things much more gracefully.

cc:  => patrik.martinsson@smhi.se
owner: somebody => jzeleny

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-11-02 15:32:02](https://pagure.io/SSSD/sssd/issue/1040#236673)**

Bumping the priority of this ticket to "critical", as we're seeing this issue occur more as people are using SSSD for longer periods in their environments.

priority: major => critical

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-11-02 15:34:29](https://pagure.io/SSSD/sssd/issue/1040#236674)**

Fields changed

summary: "getgrgid call returned more than one result" after group name change in MSAD => SSSD does not handle group renames safely

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-11-02 17:08:33](https://pagure.io/SSSD/sssd/issue/1040#236675)**

Fields changed

patch: 0 => 1
status: new => assigned

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-11-02 19:07:19](https://pagure.io/SSSD/sssd/issue/1040#236676)**

Fields changed

milestone: SSSD 1.7.0 => SSSD 1.6.3

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-11-02 19:12:55](https://pagure.io/SSSD/sssd/issue/1040#236677)**

Fixed by:
    - 583f7e8c7178f2019df0f00d9dafe973e88ed707 (master)
    - 42fb63b478058d3eb9a483e640958bd63114d309 (sssd-1-6)

resolution:  => fixed
status: assigned => closed

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2011-12-16 16:03:30](https://pagure.io/SSSD/sssd/issue/1040#236678)**

Fields changed

rhbz:  => [https://bugzilla.redhat.com/show_bug.cgi?id=746181 746181]

---

**Comment from [sgallagh](https://pagure.io/user/sgallagh) at [2017-02-24 14:54:20](https://pagure.io/SSSD/sssd/issue/1040#236679)**

**Metadata Update from @sgallagh**:
- Issue assigned to jzeleny
- Issue set to the milestone: SSSD 1.6.3
